src/main/java/net/minestom/server/registry/DynamicRegistryImpl.java

```diff
@@ -163,23 +164,32 @@ public boolean remove(@NotNull NamespaceID namespaceId) throws UnsupportedOperat
             entryByName.remove(namespaceId);
             idByName.remove(id);
             packById.remove(id);
-            vanillaRegistryDataPacket.invalidate();
+            if (vanillaRegistryDataPacket != null) {
+                vanillaRegistryDataPacket.invalidate();
+            }
             return true;
         } finally {
             lock.unlock();
         }
     }
 
     @Override
-    public @NotNull SendablePacket registryDataPacket(boolean excludeVanilla) {
+    public @NotNull SendablePacket registryDataPacket(Registries registries, boolean excludeVanilla) {
         // We cache the vanilla packet because that is by far the most common case. If some client claims not to have
         // the vanilla datapack we can compute the entire thing.
-        return excludeVanilla ? vanillaRegistryDataPacket : createRegistryDataPacket(false);
+        if (excludeVanilla) {
```

Comment:
This feels very gross, and also doesn't handle the case where we get a different Registries instance on the second call (which is weird, but you are exposing that by adding this parameter).

Either this should also track which registries it was and invalidate if we get a different one, or just take registries when constructing it would probably be fine too.

If you keep this parameter, it needs a NotNull annotation.